File: macos/Sources/WGFileSender/Core/UpdateService.swift
import Foundation

/// Minimal subset of a GitHub release we care about.
struct GitHubRelease: Codable {
    let tagName: String
    let htmlUrl: String
    let body: String?
    let prerelease: Bool
    let assets: [Asset]

    struct Asset: Codable {
        let name: String
        let browserDownloadUrl: String
        let size: Int64
    }
}

/// A newer release resolved for this platform.
struct UpdateInfo: Equatable {
    let version: String         // normalized, e.g. "0.2.0"
    let releaseNotes: String
    let pageUrl: String
    let assetUrl: String?       // direct download for this platform; nil → open the page
    let assetName: String?
    let assetSize: Int64
}

enum UpdateState: Equatable {
    case idle
    case checking
    case upToDate
    case available(UpdateInfo)
    case downloading(Double)    // progress 0…1
    case downloaded(URL)        // local file ready to open
    case failed(String)
}

enum UpdateError: LocalizedError {
    case http(Int)
    case noAsset
    case badResponse

    var errorDescription: String? {
        switch self {
        case .http(let c): return "GitHub returned \(c)"
        case .noAsset: return "No downloadable asset for this platform"
        case .badResponse: return "Unexpected response"
        }
    }
}

/// Checks GitHub Releases for a newer build and downloads the macOS asset.
final class UpdateService {
    static let repo = "bsiku3622/WGFileSender"

    /// This build's version, read from the bundle (falls back to 0.0.0 in dev runs).
    let currentVersion: String

    init() {
        currentVersion = (Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String) ?? "0.0.0"
    }

    /// Returns update details if the latest release is newer than the running build, else nil.
    func checkForUpdate() async throws -> UpdateInfo? {
        let release = try await latestRelease()
        let latest = Self.normalize(release.tagName)
        guard Self.isNewer(latest, than: currentVersion) else { return nil }
        let asset = Self.macAsset(release.assets)
        return UpdateInfo(version: latest, releaseNotes: (release.body ?? "").trimmingCharacters(in: .whitespacesAndNewlines),
                          pageUrl: release.htmlUrl, assetUrl: asset?.browserDownloadUrl,
                          assetName: asset?.name, assetSize: asset?.size ?? 0)
    }

    /// Downloads the asset into ~/Downloads and returns its local URL.
    func download(_ info: UpdateInfo, progress: @escaping (Double) -> Void) async throws -> URL {
        guard let urlStr = info.assetUrl, let url = URL(string: urlStr) else { throw UpdateError.noAsset }
        let delegate = DownloadProgressDelegate(onProgress: progress)
        let (tempURL, resp) = try await URLSession.shared.download(from: url, delegate: delegate)
        if let http = resp as? HTTPURLResponse, http.statusCode != 200 { throw UpdateError.http(http.statusCode) }
        let downloads = FileManager.default.urls(for: .downloadsDirectory, in: .userDomainMask)[0]
        let dest = downloads.appendingPathComponent(info.assetName ?? url.lastPathComponent)
        try? FileManager.default.removeItem(at: dest)
        try FileManager.default.moveItem(at: tempURL, to: dest)
        return dest
    }

    // MARK: networking

    private func latestRelease() async throws -> GitHubRelease {
        var req = URLRequest(url: URL(string: "https://api.github.com/repos/\(Self.repo)/releases/latest")!)
        req.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        req.timeoutInterval = 20
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw UpdateError.badResponse }
        guard http.statusCode == 200 else { throw UpdateError.http(http.statusCode) }
        let dec = JSONDecoder()
        dec.keyDecodingStrategy = .convertFromSnakeCase
        return try dec.decode(GitHubRelease.self, from: data)
    }

    // MARK: asset selection + version compare

    /// Prefers a .dmg, then a mac-tagged .zip, then any .zip.
    static func macAsset(_ assets: [GitHubRelease.Asset]) -> GitHubRelease.Asset? {
        assets.first { $0.name.lowercased().hasSuffix(".dmg") }
            ?? assets.first { let n = $0.name.lowercased(); return n.contains("mac") && n.hasSuffix(".zip") }
            ?? assets.first { $0.name.lowercased().hasSuffix(".zip") }
    }

    static func normalize(_ tag: String) -> String {
        tag.hasPrefix("v") ? String(tag.dropFirst()) : tag
    }

    static func isNewer(_ a: String, than b: String) -> Bool { compare(a, b) > 0 }

    /// semver-ish: numeric major.minor.patch; a pre-release (`-…`) ranks below its release.
    static func compare(_ a: String, _ b: String) -> Int {
        let (av, ap) = splitPre(a)
        let (bv, bp) = splitPre(b)
        for i in 0..<3 where av[i] != bv[i] { return av[i] < bv[i] ? -1 : 1 }
        if ap.isEmpty != bp.isEmpty { return ap.isEmpty ? 1 : -1 }
        if ap != bp { return ap < bp ? -1 : 1 }
        return 0
    }

    private static func splitPre(_ s: String) -> ([Int], String) {
        let parts = s.split(separator: "-", maxSplits: 1)
        var nums = (parts.first.map(String.init) ?? "0").split(separator: ".").map { Int($0) ?? 0 }
        while nums.count < 3 { nums.append(0) }
        return (Array(nums.prefix(3)), parts.count > 1 ? String(parts[1]) : "")
    }
}

/// Reports download progress via the session delegate's didWriteData.
final class DownloadProgressDelegate: NSObject, URLSessionDownloadDelegate {
    private let onProgress: (Double) -> Void
    init(onProgress: @escaping (Double) -> Void) { self.onProgress = onProgress }

    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask,
                    didFinishDownloadingTo location: URL) {}

    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask,
                    didWriteData bytesWritten: Int64, totalBytesWritten: Int64,
                    totalBytesExpectedToWrite: Int64) {
        guard totalBytesExpectedToWrite > 0 else { return }
        onProgress(Double(totalBytesWritten) / Double(totalBytesExpectedToWrite))
    }
}

File: android/app/src/main/java/com/jaewonbaek/wgfilesender/net/UpdateService.kt
package com.jaewonbaek.wgfilesender.net

import android.content.Context
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.engine.cio.CIO
import io.ktor.client.plugins.HttpTimeout
import io.ktor.client.plugins.onDownload
import io.ktor.client.plugins.timeout
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.statement.bodyAsText
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.io.File

@Serializable
data class GitHubRelease(
    @SerialName("tag_name") val tagName: String,
    @SerialName("html_url") val htmlUrl: String,
    val body: String? = null,
    val prerelease: Boolean = false,
    val assets: List<ReleaseAsset> = emptyList()
)

@Serializable
data class ReleaseAsset(
    val name: String,
    @SerialName("browser_download_url") val browserDownloadUrl: String,
    val size: Long = 0
)

/** A newer release resolved for this platform. */
data class UpdateInfo(
    val version: String,
    val releaseNotes: String,
    val pageUrl: String,
    val assetUrl: String?,   // direct .apk download; null → open the page
    val assetName: String?,
    val assetSize: Long
)

sealed interface UpdateState {
    data object Idle : UpdateState
    data object Checking : UpdateState
    data object UpToDate : UpdateState
    data class Available(val info: UpdateInfo) : UpdateState
    data class Downloading(val progress: Float) : UpdateState
    data class Downloaded(val file: File) : UpdateState
    data class Failed(val message: String) : UpdateState
}

/** Checks GitHub Releases for a newer build and downloads the APK. */
class UpdateService(private val context: Context) {
    private val json = Json { ignoreUnknownKeys = true }
    private val client = HttpClient(CIO) { install(HttpTimeout) }

    /** This build's versionName (falls back to 0.0.0 if unreadable). */
    val currentVersion: String = runCatching {
        context.packageManager.getPackageInfo(context.packageName, 0).versionName
    }.getOrNull() ?: "0.0.0"

    /** Update details if the latest release is newer than the running build, else null. */
    suspend fun checkForUpdate(): UpdateInfo? {
        val release = latestRelease()
        val latest = normalize(release.tagName)
        if (!isNewer(latest, currentVersion)) return null
        val asset = apkAsset(release.assets)
        return UpdateInfo(
            version = latest,
            releaseNotes = release.body?.trim().orEmpty(),
            pageUrl = release.htmlUrl,
            assetUrl = asset?.browserDownloadUrl,
            assetName = asset?.name,
            assetSize = asset?.size ?: 0L
        )
    }

    /** Downloads the APK into the cache dir and returns the file. */
    suspend fun download(info: UpdateInfo, onProgress: (Float) -> Unit): File {
        val url = info.assetUrl ?: error("no asset")
        val bytes: ByteArray = client.get(url) {
            onDownload { sent, total ->
                val t = total ?: info.assetSize
                if (t > 0) onProgress((sent.toFloat() / t).coerceIn(0f, 1f))
            }
        }.body()
        val file = File(context.cacheDir, info.assetName ?: "update.apk")
        file.writeBytes(bytes)
        return file
    }

    private suspend fun latestRelease(): GitHubRelease {
        val resp = client.get("https://api.github.com/repos/$REPO/releases/latest") {
            header("Accept", "application/vnd.github+json")
            timeout { requestTimeoutMillis = 20_000 }
        }
        check(resp.status.value == 200) { "github ${resp.status.value}" }
        return json.decodeFromString(resp.bodyAsText())
    }

    companion object {
        const val REPO = "bsiku3622/WGFileSender"

        fun normalize(tag: String): String = tag.removePrefix("v")

        fun apkAsset(assets: List<ReleaseAsset>): ReleaseAsset? =
            assets.firstOrNull { it.name.lowercase().endsWith(".apk") }

        fun isNewer(a: String, b: String): Boolean = compareVer(a, b) > 0

        /** semver-ish: numeric major.minor.patch; a pre-release (`-…`) ranks below its release. */
        fun compareVer(a: String, b: String): Int {
            val (av, ap) = splitPre(a)
            val (bv, bp) = splitPre(b)
            for (i in 0..2) if (av[i] != bv[i]) return if (av[i] < bv[i]) -1 else 1
            if (ap.isEmpty() != bp.isEmpty()) return if (ap.isEmpty()) 1 else -1
            return ap.compareTo(bp)
        }

        private fun splitPre(s: String): Pair<List<Int>, String> {
            val parts = s.split("-", limit = 2)
            val nums = parts[0].split(".").map { it.toIntOrNull() ?: 0 }
            val padded = (nums + listOf(0, 0, 0)).take(3)
            return padded to (parts.getOrNull(1) ?: "")
        }
    }
}
